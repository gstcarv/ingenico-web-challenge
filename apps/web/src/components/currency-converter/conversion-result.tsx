import { useConversionStore } from "../../stores/conversion-store";

export const ConversionResult = () => {
    const { conversionResult } = useConversionStore();

    if (!conversionResult) {
        return null;
    }

    return (
        <div className="mt-6 p-4 bg-success-50 border border-success-200 rounded-lg">
            <h3 className="text-lg font-semibold text-brand-700 mb-2">Conversion Result</h3>
            <div className="space-y-2">
                <div className="flex justify-between items-center">
                    <span className="text-neutral-700 font-medium">Original Amount</span>
                    <span className="text-brand-800 font-bold">
                        {conversionResult.originalAmount} {conversionResult.fromCurrency}
                    </span>
                </div>
                <div className="flex justify-between items-center">
                    <span className="text-neutral-700 font-medium">Converted Amount</span>
                    <span className="text-brand-800 font-bold">
                        {conversionResult.convertedAmount} {conversionResult.toCurrency}
                    </span>
                </div>
                <div className="flex justify-between items-center">
                    <span className="text-neutral-700 font-medium">Exchange Rate</span>
                    <span className="text-brand-800 font-bold">
                        1 {conversionResult.fromCurrency} = {conversionResult.exchangeRate.toFixed(4)}{" "}
                        {conversionResult.toCurrency}
                    </span>
                </div>
                <div className="text-sm text-neutral-600 mt-2">
                    Date: {conversionResult.date} | Last updated:{" "}
                    {new Date(conversionResult.lastUpdatedAt).toLocaleString()}
                </div>
            </div>
        </div>
    );
};
