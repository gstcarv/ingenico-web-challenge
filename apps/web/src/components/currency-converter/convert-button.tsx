import { Button } from "@ingenico-challenge/ui";
import { useState } from "react";

interface ConvertButtonProps {
    isSubmitting: boolean;
    hasError: boolean;
    onRetry: () => void;
    onSubmit: (e: React.FormEvent) => void;
}

export const ConvertButton = ({ isSubmitting, hasError, onRetry, onSubmit }: ConvertButtonProps) => {
    const [isRetrying, setIsRetrying] = useState(false);

    const handleRetry = async () => {
        setIsRetrying(true);
        try {
            await onRetry();
        } finally {
            setIsRetrying(false);
        }
    };

    if (hasError && !isRetrying) {
        return (
            <Button
                variant="destructive"
                className="w-full font-medium py-3 px-4 rounded-md transition-colors duration-200 focus:outline-none focus:ring-2 focus:ring-error-500 focus:ring-offset-2"
                type="button"
                onClick={handleRetry}
                disabled={isRetrying}
            >
                An error occurred, try again
            </Button>
        );
    }

    return (
        <Button
            className="w-full bg-blue-600 hover:bg-blue-700 text-white font-medium py-3 px-4 rounded-md transition-colors duration-200 focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2"
            type="submit"
            disabled={isSubmitting}
            onClick={onSubmit}
        >
            {isSubmitting ? "Converting..." : "Convert"}
        </Button>
    );
};
