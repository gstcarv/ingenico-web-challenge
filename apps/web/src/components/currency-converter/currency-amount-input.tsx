import { CurrencyInput } from "@ingenico-challenge/ui";
import { useQuery } from "@tanstack/react-query";
import { Controller, Control, useWatch } from "react-hook-form";
import { getAvailableCurrenciesQuery } from "../../queries/get-available-currencies";
import { CurrencyFormData } from "../../lib/helpers/currency-schema";
import { ReactNode } from "react";

interface CurrencyAmountInputProps {
    control: Control<CurrencyFormData>;
    suffix?: ReactNode;
}

export const CurrencyAmountInput = ({ control, suffix }: CurrencyAmountInputProps) => {
    const { data: availableCurrencies } = useQuery(getAvailableCurrenciesQuery());

    // Watch the fromCurrency to get the selected currency
    const fromCurrency = useWatch({
        control,
        name: "fromCurrency",
    });

    // Function to determine text size based on symbol length
    const getTextSizeClass = (symbol: string) => {
        if (!symbol) return "text-lg";

        const length = symbol.length;
        if (length <= 1) return "text-lg font-semibold";
        if (length <= 2) return "text-base font-semibold";
        if (length <= 3) return "text-sm font-medium";
        return "text-xs font-medium";
    };

    return (
        <Controller
            name="amount"
            control={control}
            render={({ field }) => {
                const selectedCurrency = availableCurrencies?.data?.[fromCurrency];
                const symbol = selectedCurrency?.symbol_native || selectedCurrency?.symbol || "";
                const textSizeClass = getTextSizeClass(symbol);

                return (
                    <CurrencyInput
                        id="amount"
                        value={field.value}
                        onChange={field.onChange}
                        className="w-full shadow-none border-gray-200 focus:border-blue-500 focus:ring-1 focus:ring-blue-500"
                        prefix={symbol && <span className={`${textSizeClass} text-gray-700`}>{symbol}</span>}
                        suffix={suffix}
                    />
                );
            }}
        />
    );
};
