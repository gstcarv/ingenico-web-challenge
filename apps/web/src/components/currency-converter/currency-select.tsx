import { Field, Select } from "@ingenico-challenge/ui";
import { useQuery } from "@tanstack/react-query";
import { getAvailableCurrenciesQuery } from "../../queries/get-available-currencies";
import { Controller, type Control } from "react-hook-form";
import type { CurrencyFormData } from "../../lib/helpers/currency-schema";

interface CurrencySelectProps {
    label: string;
    name: keyof CurrencyFormData;
    control: Control<CurrencyFormData>;
    className?: string;
    error?: string;
}

function getFlag(code: string) {
    return `https://raw.githubusercontent.com/Lissy93/currency-flags/refs/heads/master/assets/flags_svg/${code.toLowerCase()}.svg`;
}

export const CurrencySelect = ({ label, name, control, className, error }: CurrencySelectProps) => {
    const availableCurrencies = useQuery(getAvailableCurrenciesQuery());

    const currencies = availableCurrencies.data?.data || {};

    return (
        <Field className="flex-1 w-full">
            <Field.Label htmlFor={name}>{label}</Field.Label>
            <Controller
                name={name}
                control={control}
                render={({ field }) => {
                    return (
                        <Select
                            id={name}
                            value={field.value as string}
                            onChange={field.onChange}
                            className={`w-full shadow-none border-gray-200 focus:border-blue-500 focus:ring-1 focus:ring-blue-500 ${className || ""}`}
                            prefix={
                                <div
                                    className="w-6 h-6 rounded-full bg-gray-100 bg-neutral-50"
                                    style={{
                                        backgroundImage: `url(${getFlag(field.value as string)})`,
                                        backgroundSize: "cover",
                                        backgroundPosition: "center",
                                    }}
                                ></div>
                            }
                        >
                            <Select.Option disabled value="" selected>
                                Select a currency
                            </Select.Option>

                            {Object.entries(currencies).map(([code, currency]) => (
                                <Select.Option key={code} value={code}>
                                    {code} - {currency.name}
                                </Select.Option>
                            ))}
                        </Select>
                    );
                }}
            />
            {error && <Field.Validation>{error}</Field.Validation>}
        </Field>
    );
};
