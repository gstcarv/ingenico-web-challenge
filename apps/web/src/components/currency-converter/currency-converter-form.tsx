import { Button, DatePicker, Field } from "@ingenico-challenge/ui";
import { ArrowLeftRight, ArrowUpDown, Calendar } from "lucide-react";
import { CurrencySelect } from "./currency-select";
import { CurrencyAmountInput } from "./currency-amount-input";
import { ConvertButton } from "./convert-button";
import { useCurrencyForm } from "../../hooks/use-currency-form";
import { Controller } from "react-hook-form";

export const CurrencyConverterForm = () => {
    const { form, onSubmit, handleSwapCurrencies, handleRetry, isSubmitting, errors, conversionError } =
        useCurrencyForm();

    return (
        <form onSubmit={onSubmit} className="flex flex-col gap-6">
            <div className="flex flex-col md:flex-row items-center justify-center gap-1 md:gap-6">
                <CurrencySelect
                    label="From"
                    name="fromCurrency"
                    control={form.control}
                    error={errors.fromCurrency?.message}
                />

                <Button
                    variant="outline"
                    size="sm"
                    className="w-10 h-10 rounded-full mt-5"
                    type="button"
                    onClick={handleSwapCurrencies}
                    aria-label="Swap currencies"
                >
                    <ArrowLeftRight className="hidden lg:block" aria-hidden="true" />
                    <ArrowUpDown className="lg:hidden" aria-hidden="true" />
                </Button>

                <CurrencySelect
                    label="To"
                    name="toCurrency"
                    control={form.control}
                    error={errors.toCurrency?.message}
                />
            </div>

            <Field>
                <Field.Label htmlFor="amount">Value</Field.Label>
                <CurrencyAmountInput
                    control={form.control}
                    suffix={
                        <Controller
                            name="date"
                            control={form.control}
                            render={({ field: dateField }) => (
                                <DatePicker
                                    value={dateField.value}
                                    onChange={dateField.onChange}
                                    renderInput={({ onClick, value }) => (
                                        <Button variant="outline" size="sm" type="button" onClick={onClick}>
                                            <Calendar className="size-4 mr-2" /> {value || "Select a date"}
                                        </Button>
                                    )}
                                />
                            )}
                        />
                    }
                />

                {errors.amount && <Field.Validation>{errors.amount.message}</Field.Validation>}
            </Field>

            <div className="pt-2">
                <ConvertButton
                    isSubmitting={isSubmitting}
                    hasError={!!conversionError}
                    onRetry={handleRetry}
                    onSubmit={onSubmit}
                />
            </div>
        </form>
    );
};
