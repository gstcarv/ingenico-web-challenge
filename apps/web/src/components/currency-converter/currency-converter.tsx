import { CurrencyConverterForm } from "./currency-converter-form";
import { ConversionResult } from "./conversion-result";

export const CurrencyConverter = () => {
    return (
        <div className="bg-white w-full lg:w-[850px] rounded-lg p-8 shadow-md transition-[height]">
            <CurrencyConverterForm />
            <ConversionResult />
        </div>
    );
};
