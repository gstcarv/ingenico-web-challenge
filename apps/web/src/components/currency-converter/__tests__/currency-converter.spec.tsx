import { render, screen } from "@testing-library/react";
import { describe, expect, it } from "vitest";
import { CurrencyConverter } from "../currency-converter";

// Mock child components
vi.mock("../currency-converter-form", () => ({
    CurrencyConverterForm: () => <div data-testid="currency-converter-form">Currency Converter Form</div>,
}));

vi.mock("../conversion-result", () => ({
    ConversionResult: () => <div data-testid="conversion-result">Conversion Result</div>,
}));

describe("CurrencyConverter", () => {
    it("should render the component with correct structure", () => {
        render(<CurrencyConverter />);

        const container = screen.getByTestId("currency-converter-form").parentElement;
        expect(container).toBeInTheDocument();
        expect(container).toHaveClass(
            "bg-white",
            "w-full",
            "lg:w-[850px]",
            "rounded-lg",
            "p-8",
            "shadow-md",
            "transition-[height]",
        );
    });

    it("should render CurrencyConverterForm component", () => {
        render(<CurrencyConverter />);

        expect(screen.getByTestId("currency-converter-form")).toBeInTheDocument();
        expect(screen.getByText("Currency Converter Form")).toBeInTheDocument();
    });

    it("should render ConversionResult component", () => {
        render(<CurrencyConverter />);

        expect(screen.getByTestId("conversion-result")).toBeInTheDocument();
        expect(screen.getByText("Conversion Result")).toBeInTheDocument();
    });

    it("should have correct layout structure", () => {
        const { container } = render(<CurrencyConverter />);

        const mainDiv = container.firstChild as HTMLElement;
        expect(mainDiv.tagName).toBe("DIV");
        expect(mainDiv.children).toHaveLength(2);
    });

    it("should match snapshot", () => {
        const { container } = render(<CurrencyConverter />);
        expect(container.firstChild).toMatchSnapshot();
    });
});
