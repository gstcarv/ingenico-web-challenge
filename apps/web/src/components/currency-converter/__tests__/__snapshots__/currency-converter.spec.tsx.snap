// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CurrencyConverter > should match snapshot 1`] = `
<div
  class="bg-white w-full lg:w-[850px] rounded-lg p-8 shadow-md transition-[height]"
>
  <div
    data-testid="currency-converter-form"
  >
    Currency Converter Form
  </div>
  <div
    data-testid="conversion-result"
  >
    Conversion Result
  </div>
</div>
`;
